frameworks.qmd

```diff
@@ -1,119 +1,1793 @@
+patterns. Dropout for instance randomly sets a fraction of input units
+to 0 at each update during training, which helps prevent overfitting.
+
+Frameworks also provide efficient implementations of gradient descent,
+Adagrad, Adadelta, and Adam. Adding regularization like dropout and
+L1/L2 penalties prevents overfitting during training. Batch
+normalization accelerates training by normalizing inputs to layers.
+
+### Model Training Support
+
+Before training a defined neural network model, a compilation step is
+required. During this step, the high-level architecture of the neural
+network is transformed into an optimized, executable format. This
+process comprises several steps. The construction of the computational
+graph is the first step. It represents all the mathematical operations
+and data flow within the model. We discussed this earlier.
+
+During training, the focus is on executing the computational graph.
+Every parameter within the graph, such as weights and biases, is
+assigned an initial value. This value might be random or based on a
+predefined logic, depending on the chosen initialization method.
+
+The next critical step is memory allocation. Essential memory is
+reserved for the model\'s operations on both CPUs and GPUs, ensuring
+efficient data processing. The model\'s operations are then mapped to
+the available hardware resources, particularly GPUs or TPUs, to expedite
+computation. Once compilation is finalized, the model is prepared for
+training.
+
+The training process employs various tools to enhance efficiency. Batch
+processing is commonly used to maximize computational throughput.
+Techniques like vectorization enable operations on entire data arrays,
+rather than proceeding element-wise, which bolsters speed. Optimizations
+such as kernel fusion (refer to the Optimizations chapter) amalgamate
+multiple operations into a single action, minimizing computational
+overhead. Operations can also be segmented into phases, facilitating the
+concurrent processing of different mini-batches at various stages.
+
+Frameworks consistently checkpoint the state, preserving intermediate
+model versions during training. This ensures that if an interruption
+occurs, the progress isn\'t wholly lost, and training can recommence
+from the last checkpoint. Additionally, the system vigilantly monitors
+the model\'s performance against a validation data set. Should the model
+begin to overfit (that is, if its performance on the validation set
+declines), training is automatically halted, conserving computational
+resources and time.
+
+ML frameworks incorporate a blend of model compilation, enhanced batch
+processing methods, and utilities such as checkpointing and early
+stopping. These resources manage the complex aspects of performance,
+enabling practitioners to zero in on model development and training. As
+a result, developers experience both speed and ease when utilizing the
+capabilities of neural networks.
```

Comment:
It would be helpful to have a diagram of batch processing here so the reader can visualize how batch processing works. For example: https://media.geeksforgeeks.org/wp-content/uploads/20221020142629/Batchoperatingsystem1-660x330.png